src/DiffTests.jl

```diff
@@ -114,16 +114,16 @@ end
 
 self_weighted_logit(x) = inv(1.0 + exp(-dot(x, x)))
 
-nested_array_mul(x) = sum(sum(x[1] * [[x[2], x[3]]]))
+nested_array_mul(x)::eltype(x) = sum(sum(x[1] * [[x[2], x[3]]]))
```

Comment:
The type annotation is wrong in general (one example is `x::Array{Bool}`).
```suggestion
nested_array_mul(x) = sum(sum(x[1] * [[x[2], x[3]]]))
```

Reply:
Yes, you are right, that's the shakiest part of the PR (this particular one could be fixed, but there is no single contract for the whole family of functions). My motivation was that if all functions are type stable, that might help with the testing times in ForwardDiff and especially ReverseDiff, which are abysmal.
Do you think it's a good thing to enforce type stability for these functions? How would you approach that?

Reply:
They should be type stable if `sum` etc. are implemented in a type-stable way. But I guess maybe you rather mean that the type of the argument and the return type are the same/related?

Reply:
Generally, I think the best is to remove all these return type annotations and first come up with concrete problems. Then it is also possible to figure out whether these problems should be fixed in DiffTests or e.g. in the AD packages.

Reply:
without this type annotation the `@inferred` macro, which I added to the tests, fails. IIUC for this function it is inferred to Any. I can drop these checks, or I can correct the type annotation, or adjust the function definition to be type stable. One real problem is the test times, so type stability may help a bit with that.

Reply:
I'd suggest removing the `@inferred` statements and possibly opening an issue. I also don't know if there's much we can do here since these issues are caused by type inference problems in Base. On Julia 1.9.1 I get:
```julia
julia> using Test

julia> f(x) = x[1] * [[x[2], x[3]]]
f (generic function with 1 method)

julia> x = rand(3)
3-element Vector{Float64}:
 0.1648801068106326
 0.14897187763070263
 0.3251905884653816

julia> @inferred(f(x))
ERROR: return type Vector{Vector{Float64}} does not match inferred return type Any
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ REPL[20]:1
```

Reply:
I have fixed the type inference of `nested_array_mul()`, it should be good for Bool-valued input vectors (however, only Float64 is tested at the moment, and some functions (mutating ones) are not compatible with Bool-valued vectors).